# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [33]:
# Import pandas and load dfs

import pandas as pd

# Load customer data dataset an from an insurance company
url1 = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv'
customer_df1 = pd.read_csv(url1)
customer_df1.head(2)

# Load file 2 from an insurance company
url2 = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv'
customer_df2 = pd.read_csv(url2)
customer_df2.head(2)

# Load file 3 from an insurance company
url3 = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv'
customer_df3 = pd.read_csv(url3)
customer_df3.head(2)

# Define how floats are formatted when displayed with the string formatting function: '{:,.2f}'.format
# , → adds a thousands separator (e.g., 1,000 instead of 1000)
# .2f → rounds numbers to two decimal places
pd.options.display.float_format = '{:,.2f}'.format

In [34]:
customer_df1.head(2)

,Customer,ST,GENDER,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount
0,RB50392,Washington,NaN,Master,NaN,0.00,"1,000.00",1/0/00,Personal Auto,Four-Door Car,2.70
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.00,94.00,1/0/00,Personal Auto,Four-Door Car,"1,131.46"


In [35]:
customer_df2.head(2)

,Customer,ST,GENDER,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Total Claim Amount,Policy Type,Vehicle Class
0,GS98873,Arizona,F,Bachelor,323912.47%,16061,88,1/0/00,633.60,Personal Auto,Four-Door Car
1,CW49887,California,F,Master,462680.11%,79487,114,1/0/00,547.20,Special Auto,SUV


In [36]:
customer_df3.head(2)

,Customer,State,Customer Lifetime Value,Education,Gender,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Total Claim Amount,Vehicle Class
0,SA25987,Washington,"3,479.14",High School or Below,M,0,104,0,Personal Auto,499.20,Two-Door Car
1,TB86706,Arizona,"2,502.64",Master,M,0,66,0,Personal Auto,3.47,Two-Door Car


#### Check shape, types and nulls of each df
 First combine the three datasets and then apply the cleaning function to the new combined dataset

In [37]:
customer_df1.shape

(4008, 11)

In [38]:
#Check the type of the variables of the df
customer_df1.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

In [39]:
customer_df1.isna().sum()

Customer                     2937
ST                           2937
GENDER                       3054
Education                    2937
Customer Lifetime Value      2940
Income                       2937
Monthly Premium Auto         2937
Number of Open Complaints    2937
Policy Type                  2937
Vehicle Class                2937
Total Claim Amount           2937
dtype: int64

In [40]:
customer_df2.shape

(996, 11)

In [41]:
#Check the type of the variables of the df
customer_df2.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                         int64
Monthly Premium Auto           int64
Number of Open Complaints     object
Total Claim Amount           float64
Policy Type                   object
Vehicle Class                 object
dtype: object

In [42]:
customer_df2.isna().sum()

Customer                     0
ST                           0
GENDER                       5
Education                    0
Customer Lifetime Value      4
Income                       0
Monthly Premium Auto         0
Number of Open Complaints    0
Total Claim Amount           0
Policy Type                  0
Vehicle Class                0
dtype: int64

In [43]:
customer_df3.shape

(7070, 11)

In [44]:
# Check the type of the variables of the df
customer_df3.dtypes

Customer                      object
State                         object
Customer Lifetime Value      float64
Education                     object
Gender                        object
Income                         int64
Monthly Premium Auto           int64
Number of Open Complaints      int64
Policy Type                   object
Total Claim Amount           float64
Vehicle Class                 object
dtype: object

In [45]:
customer_df3.isna().sum()

Customer                     0
State                        0
Customer Lifetime Value      0
Education                    0
Gender                       0
Income                       0
Monthly Premium Auto         0
Number of Open Complaints    0
Policy Type                  0
Total Claim Amount           0
Vehicle Class                0
dtype: int64

#### Prepare the df for the concatenation by standardise col names and format the data

In [46]:
# Normalize text columns (strip spaces, set lowercase)

# Parameters:
# df : DataFrame
# cols : columns to normalize; if None, all object/string columns
# lower : convert to lowercase
# strip : strip leading/trailing whitespace
# normalize_columns: normalize column names (strip, lower, replace spaces with _)

# Returns:
# DataFrame (same object, modified in place style but returns df for chaining)


from typing import Optional, Iterable
def standardize_text(
    df: pd.DataFrame,
    cols: Optional[Iterable[str]] = None,
    lower: bool = True,
    strip: bool = True,
    normalize_columns: bool = True,
) -> pd.DataFrame:
       
# Standardizing string values (inside the cells)
    if cols is None:
        cols = df.select_dtypes(include=["object", "string"]).columns  # all object/string columns
    for c in cols:
        s = df[c].astype("string")  # ensure string dtype
        if strip:
            s = s.str.strip()  # removes leading/trailing spaces
        if lower:
            s = s.str.lower()  # convert to lowercase
        df[c] = s

# Standardizing column names
    if normalize_columns:
        df.columns = (
            df.columns
            .str.strip()
            .str.lower()
            .str.replace(" ", "_")
        )
    return df

customer_df1 = standardize_text(customer_df1)
customer_df2 = standardize_text(customer_df2)
customer_df3 = standardize_text(customer_df3)

In [47]:
#Check the type of the variables of the df
customer_df1.dtypes

customer                     string[python]
st                           string[python]
gender                       string[python]
education                    string[python]
customer_lifetime_value      string[python]
income                              float64
monthly_premium_auto                float64
number_of_open_complaints    string[python]
policy_type                  string[python]
vehicle_class                string[python]
total_claim_amount                  float64
dtype: object

In [48]:
#Check the type of the variables of the df
customer_df2.dtypes

customer                     string[python]
st                           string[python]
gender                       string[python]
education                    string[python]
customer_lifetime_value      string[python]
income                                int64
monthly_premium_auto                  int64
number_of_open_complaints    string[python]
total_claim_amount                  float64
policy_type                  string[python]
vehicle_class                string[python]
dtype: object

In [49]:
# Check the type of the variables of the df
customer_df3.dtypes

customer                     string[python]
state                        string[python]
customer_lifetime_value             float64
education                    string[python]
gender                       string[python]
income                                int64
monthly_premium_auto                  int64
number_of_open_complaints             int64
policy_type                  string[python]
total_claim_amount                  float64
vehicle_class                string[python]
dtype: object

In [50]:
# Standardise the column names: `st` is replaced for `state`
# Ensure that col names are the same between dfs

customer_df1.rename(columns={"st": "state"}, inplace = True)
customer_df1.head(2)

,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount
0,rb50392,washington,<NA>,master,<NA>,0.00,"1,000.00",1/0/00,personal auto,four-door car,2.70
1,qz44356,arizona,f,bachelor,697953.59%,0.00,94.00,1/0/00,personal auto,four-door car,"1,131.46"


In [51]:
# Standardise the column names: `st` is replaced for `state`

customer_df2.rename(columns={"st": "state"}, inplace = True)
customer_df2.head(2)


,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,total_claim_amount,policy_type,vehicle_class
0,gs98873,arizona,f,bachelor,323912.47%,16061,88,1/0/00,633.60,personal auto,four-door car
1,cw49887,california,f,master,462680.11%,79487,114,1/0/00,547.20,special auto,suv


### Vertical concatenation

In [52]:
# Vertical concatenation is used to add more rows;
# ignore_index=True, to create a new index for the combined df, ignoring the original indexes

customer_concat = pd.concat([customer_df1, customer_df2, customer_df3], axis=0, ignore_index=True)
customer_concat.shape


(12074, 11)

#### Clean and format the new df

#### Formatting data types

In [53]:
# Format data types of "customer lifetime value" that should be numeric
# cleansing inconsistent values of "customer_lifetime_value"

customer_concat["customer_lifetime_value"] = (
    customer_concat["customer_lifetime_value"]
    .astype("string")                       # ensure string dtype
    .fillna("")                             # temporarily replace <NA> with empty string
    .str.replace('%', '', regex=False)      # remove percent signs
    .str.strip()                            # remove leading/trailing spaces
)

# Convert to numeric safely; non-convertible values become NaN
customer_concat["customer_lifetime_value"] = pd.to_numeric(
    customer_concat["customer_lifetime_value"],
    errors="coerce"
)
customer_concat["customer_lifetime_value"].dtype
customer_concat.dtypes

customer                     string[python]
state                        string[python]
gender                       string[python]
education                    string[python]
customer_lifetime_value             Float64
income                              float64
monthly_premium_auto                float64
number_of_open_complaints            object
policy_type                  string[python]
vehicle_class                string[python]
total_claim_amount                  float64
dtype: object

In [54]:
customer_concat["customer_lifetime_value"].describe
customer_concat.isna().sum()

customer                     2937
state                        2937
gender                       3059
education                    2937
customer_lifetime_value      2944
income                       2937
monthly_premium_auto         2937
number_of_open_complaints    2937
policy_type                  2937
vehicle_class                2937
total_claim_amount           2937
dtype: int64

In [55]:
# "number of open complaints" has an incorrect format. 
# Look at the different values it takes with `unique()` and take the middle value. 
customer_concat["number_of_open_complaints"].unique()

# As an example, 1/5/00 should be 5. Number of open complaints is a string
# you can use `split()` to deal with it and take the number you need. 
# Finally, since it should be numeric, cast the column to be in its proper type.
# Ensure the column is a string
customer_concat["number_of_open_complaints"] = customer_concat["number_of_open_complaints"].astype("string")

# Split by '/' and take the middle value (index 1)
customer_concat["number_of_open_complaints"] = customer_concat["number_of_open_complaints"].str.split('/').str[1]

# Convert to numeric (Int64 allows for missing values)
customer_concat["number_of_open_complaints"] = pd.to_numeric(
    customer_concat["number_of_open_complaints"],
    errors="coerce"
).astype("Int64")

# Check result
print(customer_concat["number_of_open_complaints"].unique())
print(customer_concat["number_of_open_complaints"].dtype)





<IntegerArray>
[0, 2, 1, 3, 5, 4, <NA>]
Length: 7, dtype: Int64
Int64


In [56]:
customer_concat["customer_lifetime_value"].describe
customer_concat.isna().sum()

customer                      2937
state                         2937
gender                        3059
education                     2937
customer_lifetime_value       2944
income                        2937
monthly_premium_auto          2937
number_of_open_complaints    10007
policy_type                   2937
vehicle_class                 2937
total_claim_amount            2937
dtype: int64

In [57]:
# Check categories of nominal var "state"
customer_concat["state"].drop_duplicates() 

0       washington
1          arizona
2           nevada
3       california
5           oregon
17            cali
28              az
65              wa
1071          <NA>
Name: state, dtype: string

In [58]:
# Check categories of nominal var "gender"
customer_concat["gender"].drop_duplicates()

0        <NA>
1           f
3           m
154     femal
231      male
297    female
Name: gender, dtype: string

In [59]:
# Check categories of nominal var "education"
customer_concat["education"].drop_duplicates()

0                     master
1                   bachelor
4       high school or below
6                    college
21                 bachelors
44                    doctor
1071                    <NA>
Name: education, dtype: string

In [60]:
# Check categories of nominal var "vehicle_class"
customer_concat["vehicle_class"].drop_duplicates()

0       four-door car
2        two-door car
3                 suv
31         luxury suv
43         sports car
79         luxury car
1071             <NA>
Name: vehicle_class, dtype: string

In [61]:
# Cleansing inconsistent values (cleaning inconsistent names of categories) of "state"
customer_concat["state"] = customer_concat["state"].map({"cali": "california", "az": "arizona", "wa": "washington"}).fillna(customer_concat["state"])

# Cleansing inconsistent values (cleaning inconsistent names of categories) of "gender"
customer_concat["gender"] = customer_concat["gender"].map({"F": "f", "femal": "f", "female": "f", "male": "m"}).fillna(customer_concat["gender"])

# Cleansing inconsistent values (cleaning inconsistent names of categories) of "education"
customer_concat["education"] = customer_concat["education"].map({"bachelors": "bachelor"}).fillna(customer_concat["education"])

# Cleansing inconsistent values (cleaning inconsistent names of categories) of "vehicle_class"
customer_concat["vehicle_class"] = customer_concat["vehicle_class"].map({"sports car": "luxury", "luxury suv": "luxury", "luxury car": "luxury",}).fillna(customer_concat["vehicle_class"])

In [62]:
# Check categories of nominal var "state" after cleansing
customer_concat["state"].drop_duplicates() 

0       washington
1          arizona
2           nevada
3       california
5           oregon
1071          <NA>
Name: state, dtype: object

In [63]:
# Check categories of nominal var "gender" after cleansing
customer_concat["gender"].drop_duplicates()

0    <NA>
1       f
3       m
Name: gender, dtype: object

In [64]:
# Check categories of nominal var "education" after cleansing
customer_concat["education"].drop_duplicates()

0                     master
1                   bachelor
4       high school or below
6                    college
44                    doctor
1071                    <NA>
Name: education, dtype: object

In [65]:
# Check categories of nominal var "vehicle_class" after cleansing
customer_concat["vehicle_class"].drop_duplicates()

0       four-door car
2        two-door car
3                 suv
31             luxury
1071             <NA>
Name: vehicle_class, dtype: object

# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [66]:
import pandas as pd

# # Load file marketing_customer_analysis.csv from an insurance company
url4 = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv'
df_marketing = pd.read_csv(url4, index_col=0) # first column is an index

# Define how floats are formatted when displayed with the string formatting function: '{:,.2f}'.format
# , → adds a thousands separator (e.g., 1,000 instead of 1000)
# .2f → rounds numbers to two decimal places
pd.options.display.float_format = '{:,.2f}'.format

df_marketing.head(4)

,customer,state,customer_lifetime_value,response,coverage,education,effective_to_date,employmentstatus,gender,income,...,number_of_policies,policy_type,policy,renew_offer_type,sales_channel,total_claim_amount,vehicle_class,vehicle_size,vehicle_type,month
unnamed:_0,,,,,,,,,,,,,,,,,,,,,
0,DK49336,Arizona,"4,809.22",No,Basic,College,2011-02-18,Employed,M,48029,...,9,Corporate Auto,Corporate L3,Offer3,Agent,292.80,Four-Door Car,Medsize,A,2
1,KX64629,California,"2,228.53",No,Basic,College,2011-01-18,Unemployed,F,0,...,1,Personal Auto,Personal L3,Offer4,Call Center,744.92,Four-Door Car,Medsize,A,1
2,LZ68649,Washington,"14,947.92",No,Basic,Bachelor,2011-02-10,Employed,M,22139,...,2,Personal Auto,Personal L3,Offer3,Call Center,480.00,SUV,Medsize,A,2
3,XL78013,Oregon,"22,332.44",Yes,Extended,College,2011-01-11,Employed,M,49078,...,2,Corporate Auto,Corporate L3,Offer2,Branch,484.01,Four-Door Car,Medsize,A,1


In [67]:
# Normalize text columns (strip spaces, set lowercase)

# Parameters:
# df : DataFrame
# cols : columns to normalize; if None, all object/string columns
# lower : convert to lowercase
# strip : strip leading/trailing whitespace
# normalize_columns: normalize column names (strip, lower, replace spaces with _)

# Returns:
# DataFrame (same object, modified in place style but returns df for chaining)


from typing import Optional, Iterable
def standardize_text(
    df: pd.DataFrame,
    cols: Optional[Iterable[str]] = None,
    lower: bool = True,
    strip: bool = True,
    normalize_columns: bool = True,
) -> pd.DataFrame:
       
# Standardizing string values (inside the cells)
    if cols is None:
        cols = df.select_dtypes(include=["object", "string"]).columns  # all object/string columns
    for c in cols:
        s = df[c].astype("string")  # ensure string dtype
        if strip:
            s = s.str.strip()  # removes leading/trailing spaces
        if lower:
            s = s.str.lower()  # convert to lowercase
        df[c] = s

# Standardizing column names
    if normalize_columns:
        df.columns = (
            df.columns
            .str.strip()
            .str.lower()
            .str.replace(" ", "_")
        )
    return df

df_marketing = standardize_text(df_marketing)

In [68]:
# Convert quantitative variable from string to numeric
# If a value cannot be converted to a number, replace it with NaN (missing value): errors='coerce'
numeric_cols = ['customer_lifetime_value', 'income', 'monthly_premium_auto', 
                'months_since_last_claim', 'months_since_policy_inception',
                'number_of_open_complaints', 'number_of_policies', 'total_claim_amount']
for col in numeric_cols:
    df_marketing[col] = pd.to_numeric(df_marketing[col].astype(str).str.replace('%','', regex=False).str.strip(), errors='coerce')

In [69]:
# Extract middle value from "number_of_open_complaints" (X/Y/Z format) 
df_marketing['number_of_open_complaints'] = df_marketing['number_of_open_complaints'].astype("string").str.split('/').str[1]
df_marketing['number_of_open_complaints'] = pd.to_numeric(df_marketing['number_of_open_complaints'], errors='coerce').astype("Int64")

In [70]:
# Convert date columns
df_marketing['effective_to_date'] = pd.to_datetime(df_marketing['effective_to_date'], errors='coerce')

In [71]:
df_marketing.dtypes

customer                         string[python]
state                            string[python]
customer_lifetime_value                 float64
response                         string[python]
coverage                         string[python]
education                        string[python]
effective_to_date                datetime64[ns]
employmentstatus                 string[python]
gender                           string[python]
income                                    int64
location_code                    string[python]
marital_status                   string[python]
monthly_premium_auto                      int64
months_since_last_claim                 float64
months_since_policy_inception             int64
number_of_open_complaints                 Int64
number_of_policies                        int64
policy_type                      string[python]
policy                           string[python]
renew_offer_type                 string[python]
sales_channel                    string[

In [72]:

df_marketing.columns = df_marketing.columns.str.strip().str.lower().str.replace(' ', '_')

# List of categorical variables
categorical_col = [
    'customer',
    'state',
    'response',
    'coverage',
    'education',
    'employmentstatus',  
    'gender',
    'location_code',
    'marital_status',
    'policy_type',
    'policy',
    'renew_offer_type',
    'sales_channel',
    'vehicle_class',
    'vehicle_size',
    'vehicle_type'
]

# Keep only columns that exist in the DataFrame
categorical_col = [col for col in categorical_col if col in df_marketing.columns]

# Convert all categorical columns to string dtype
for col in categorical_col:
    df_marketing[col] = df_marketing[col].astype('string')

# Print unique values for each categorical column
for col in categorical_col:
    print(f"Unique values for '{col}':")
    print(df_marketing[col].drop_duplicates().tolist())
    print("-" * 50)

Unique values for 'customer':
['dk49336', 'kx64629', 'lz68649', 'xl78013', 'qa50777', 'ho81102', 'iw72280', 'ih64929', 'fm55990', 'qx31376', 'hg93801', 'ge63585', 'eo51920', 'kr82385', 'tl59568', 'cw49887', 'fh51383', 'tr26615', 'zc97968', 'nj54277', 'om80593', 'rl65736', 'eq62488', 'nq71171', 'yk32760', 'ay59197', 'xf15789', 'mq68407', 'ci99444', 'bd70337', 'gx19342', 'dp19210', 'oj63606', 'oh23893', 'ri76930', 'yk26558', 'oe15005', 'ra17555', 'zi34278', 'xd72019', 'rr93140', 'be25221', 'pv62583', 'bl12371', 'er49844', 'ri22468', 'sy48591', 'pr23514', 'uy32623', 'qv30663', 'ms94619', 'ft56968', 'tx39594', 'gr42174', 'xi79589', 'xq74827', 'tw17878', 'gb43363', 'br69566', 'ep83939', 'iq95180', 'jx52708', 'uf44021', 'av25133', 'wl24392', 'lo57874', 'uo54511', 'kr35099', 'ny12562', 'qg27547', 'ph78214', 'lo37570', 'fh81409', 'rw65977', 'ld75023', 'oz92426', 'fi17798', 'yc54142', 'ln28900', 'sx63814', 'vm85555', 'ro18530', 'fr48094', 'dy27653', 'oh69760', 'ch50584', 'wm19996', 'zh92587', '

In [73]:
# cleaning inconsistent values (cleaning inconsistent names of categories) of "vehicle_class"
df_marketing["vehicle_class"] = df_marketing["vehicle_class"].map({"sports car": "luxury", "luxury suv": "luxury", "luxury car": "luxury",}).fillna(df_marketing["vehicle_class"])
df_marketing["vehicle_class"].drop_duplicates()

unnamed:_0
0    four-door car
2              suv
7     two-door car
8           luxury
Name: vehicle_class, dtype: object

In [74]:
# Check NA counts for all categorical columns
for col in categorical_col:
    if col in df_marketing.columns:
        na_count = df_marketing[col].isna().sum()
        total = len(df_marketing)
        pct = (na_count / total) * 100
        print(f"{col}: {na_count} missing values ({pct:.2f}%)")

customer: 0 missing values (0.00%)
state: 0 missing values (0.00%)
response: 0 missing values (0.00%)
coverage: 0 missing values (0.00%)
education: 0 missing values (0.00%)
employmentstatus: 0 missing values (0.00%)
gender: 0 missing values (0.00%)
location_code: 0 missing values (0.00%)
marital_status: 0 missing values (0.00%)
policy_type: 0 missing values (0.00%)
policy: 0 missing values (0.00%)
renew_offer_type: 0 missing values (0.00%)
sales_channel: 0 missing values (0.00%)
vehicle_class: 0 missing values (0.00%)
vehicle_size: 0 missing values (0.00%)
vehicle_type: 0 missing values (0.00%)


In [75]:
# Change variable name to make it conform to data it contains "gender" -> "sex"
df_marketing = df_marketing.rename(columns={'gender': 'sex'})
categorical_col = [
    'customer',
    'state',
    'response',
    'coverage',
    'education',
    'employmentstatus',  
    'sex',
    'location_code',
    'marital_status',
    'policy_type',
    'policy',
    'renew_offer_type',
    'sales_channel',
    'vehicle_class',
    'vehicle_size',
    'vehicle_type'
]

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

In [76]:
# Bivariate analysis: total revenue by each sales channel
# pivot summary table showing the total revenue for each sales channel (branch, call center, web, and mail)

pivot_sales = (
    df_marketing
    .pivot_table(
        values='total_claim_amount',       # numeric column (revenue)
        index='sales_channel',             # grouping variable
        aggfunc='sum'                      # aggregate by sum of revenue
    )
    .round(2)                              # round to 2 decimal places
    .sort_values(by='total_claim_amount', ascending=False)  # sort high → low
)

# Display results
print(pivot_sales)


               total_claim_amount
sales_channel                    
agent                1,810,226.82
branch               1,301,204.00
call center            926,600.82
web                    706,600.04


In [77]:
# Ratio of total revenue from Agent to Web
ratio_agent_web = (
    pivot_sales.loc['agent', 'total_claim_amount'] /
    pivot_sales.loc['web', 'total_claim_amount']
)
print(f"Ratio of total revenue from Agent to Web: {ratio_agent_web:.2f}")


Ratio of total revenue from Agent to Web: 2.56


#### Results
- The Agent sales channel generated the highest total revenue, followed by Branch, Call Center, and Web.
- The ratio of total revenue from Agent to Web is 2.56, indicating that the Agent channel brought in about 2.6 times more revenue than Web.

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [78]:
# Bivariate analysis: average customer lifetime value by gender and education level (gender was renamed to sex)
# pivot summary table showing the average customer lifetime value by gender and education level

pivot_customer_lifetime_value = (
    df_marketing
    .pivot_table(
        values='customer_lifetime_value',   # numeric variable
        index='sex',                     # row grouping
        columns='education',                # column grouping
        aggfunc='mean'                      # average (mean)
    )
    .round(2)                               # round to 2 decimals
)

# Display results
print(pivot_customer_lifetime_value)


education  bachelor  college   doctor  high school or below   master
sex                                                                 
f          7,874.27 7,748.82 7,328.51              8,675.22 8,157.05
m          7,703.60 8,052.46 7,415.33              8,149.69 8,168.83


#### Results
- Customers with a doctorate degree show lower average customer lifetime value than those in other education categories.
- The average customer lifetime value is generally similar between male and female customers across most education levels.
- However, a notable divergence appears among customers with a high school or below education, where female customers report a higher average lifetime value compared to males.
- This suggests that sex differences in lifetime value may become more pronounced in lower education segments, while remaining relatively stable in higher education groups.

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [86]:
# "number of open complaints" has an incorrect format. 
# Look at the different values it takes with `unique()` and take the middle value. 
df_marketing["number_of_open_complaints"].unique()

# As an example, 1/5/00 should be 5. Number of open complaints is a string
# you can use `split()` to deal with it and take the number you need. 
# Finally, since it should be numeric, cast the column to be in its proper type.
# Ensure the column is a string
df_marketing["number_of_open_complaints"] = df_marketing["number_of_open_complaints"].astype("string")

# Split by '/' and take the middle value (index 1)
df_marketing["number_of_open_complaints"] = df_marketing["number_of_open_complaints"].str.split('/').str[1]

# Convert to numeric (Int64 allows for missing values)
df_marketing["number_of_open_complaints"] = pd.to_numeric(
    df_marketing["number_of_open_complaints"],
    errors="coerce"
).astype("Int64")

# Check result
print(df_marketing["number_of_open_complaints"].unique())
print(df_marketing["number_of_open_complaints"].dtype)

<IntegerArray>
[<NA>]
Length: 1, dtype: Int64
Int64


In [93]:
print(df_marketing['number_of_open_complaints'].head(10))

unnamed:_0
0    <NA>
1    <NA>
2    <NA>
3    <NA>
4    <NA>
5    <NA>
6    <NA>
7    <NA>
8    <NA>
9    <NA>
Name: number_of_open_complaints, dtype: Int64


In [94]:
# Extract month name from "effective_to_date" into a new column "month"
df_marketing['month'] = df_marketing['effective_to_date'].dt.month_name()


# Total number of complaints by policy type category over time in months
complaints_by_policy_month = (
    df_marketing
    .groupby(['policy_type', 'month'], as_index=False)
    ["number_of_open_complaints"]
    .sum()
)

# Sort for readability
complaints_by_policy_month = complaints_by_policy_month.sort_values(['policy_type', 'month'])

print(complaints_by_policy_month)


      policy_type     month  number_of_open_complaints
0  corporate auto  February                          0
1  corporate auto   January                          0
2   personal auto  February                          0
3   personal auto   January                          0
4    special auto  February                          0
5    special auto   January                          0
